util/llbutil/authprovider/cloudstoredauthprovider/tokenseed.go

```diff
@@ -0,0 +1,85 @@
+package cloudstoredauthprovider
+
+import (
+	"crypto/rand"
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"sync"
+	"syscall"
+
+	"github.com/gofrs/flock"
+	"github.com/pkg/errors"
+)
+
+type tokenSeeds struct {
+	mu  sync.Mutex
+	dir string
+	m   map[string]seed
+}
+
+type seed struct {
+	Seed []byte
+}
+
+func (ts *tokenSeeds) getSeed(host string) ([]byte, error) {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+
+	if err := os.MkdirAll(ts.dir, 0755); err != nil {
+		return nil, err
+	}
+
+	if ts.m == nil {
+		ts.m = map[string]seed{}
+	}
+
+	l := flock.New(filepath.Join(ts.dir, ".token_seed.lock"))
+	if err := l.Lock(); err != nil {
+		if !errors.Is(err, syscall.EROFS) && !errors.Is(err, os.ErrPermission) {
+			return nil, err
+		}
+	} else {
+		defer func() { _ = l.Unlock() }()
+	}
```

Comment:
I'm a little confused - if `l.Lock()` errors but the error is _not_ `syscall.EROFS` or `os.ErrPermission`, then we continue executing but never call `l.Unlock()`? Is that intentional? If so, isn't that a little risky?

Reply:
the original buildkit code simply did a `defer l.Unlock()`; however we have a linter that wants an explicit `_ = ...` to show errors are intentionally dropped, and this is the pattern we've been using to show that.

Reply:
Gah, the context here isn't showing as many lines as I'd like. What I was trying to comment on is that the `if err != nil` block does not fully stop execution on line `41`, so we can enter the `if` block due to a non-`nil` error, but still end up exiting the `if` block without entering the `else` block.

That means that we will continue execution without access to the lock (and we will never call `Unlock`). Is that both intentional and safe?

Reply:
honestly I have no idea; it comes from upstream: https://github.com/moby/buildkit/blob/master/session/auth/authprovider/tokenseed.go#L38-L44

Reply:
Ah, gotcha! At least it's an upstream thing :laughing: